aten/src/ATen/native/cuda/ForeachFunctors.cuh

```diff
@@ -245,6 +245,263 @@ struct BinaryOpListFunctor {
         }
 };
 
+template<typename T, template<class> class Op>
+struct UnaryOpFunctor_ {
+    __device__ void operator() (
+        int chunk_size,
+        TensorListMetadata<1>& tl) {
+            int tensor_loc = tl.block_to_tensor[blockIdx.x];
+            int chunk_idx = tl.block_to_chunk[blockIdx.x];
+            int n = tl.sizes[tensor_loc];
+
+            T* x = (T*)tl.addresses[0][tensor_loc];
+            x += chunk_idx * chunk_size;
+
+            n -= chunk_idx * chunk_size;
+
+            T r_x[kILP];
+
+            // to make things simple, we put aligned case in a different code path
+            if(n % kILP == 0 && chunk_size % kILP == 0 && is_aligned(x)) {
+                for(int i_start = threadIdx.x; i_start * kILP < n && i_start * kILP < chunk_size; i_start += blockDim.x) {
+                    // load
+                    load_store(r_x, x, 0 , i_start);
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_x[ii] = Op<T>()(static_cast<T>(r_x[ii]));
+                    }
+                    // store
+                    load_store(x, r_x, i_start, 0);
+                }
+            }
+            else {
+                for(int i_start = 0; i_start < n && i_start < chunk_size; i_start += blockDim.x * kILP) {
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_x[ii] = 0;
+                        int i = i_start + threadIdx.x + ii * blockDim.x;
+                        if(i < n && i < chunk_size) {
+                            r_x[ii] = x[i];
+                        }
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        r_x[ii] = Op<T>()(static_cast<T>(r_x[ii]));
+                    }
+#pragma unroll
+                    for(int ii = 0; ii < kILP; ii++) {
+                        int i = i_start + threadIdx.x + ii * blockDim.x;
+                        if(i < n && i < chunk_size)
+                            x[i] = r_x[ii];
+                    }
+                }
+            }
+        }
+};
+
+template<typename T, template<class> class Op>
+struct UnaryOpFunctor {
+    __device__ void operator() (
+        int chunk_size,
+        TensorListMetadata<2>& tl) {
```

Comment:
The structure of inplace and out of place UnaryOpFunctors seems close enough that you could unify them with another template arg and a few conditionals in the code? Especially if you get rid of r_out

Reply:
Added as a TODO, as i want to have a special PR which will refactor this whole file as 90% of the code here is repeated.